waterfall/golang/net/qemu/control.go

```diff
@@ -0,0 +1,347 @@
+func (q *ControlSocketConn) SetDeadline(t time.Time) error {
+	return errNotImplemented
+}
+
+// SetReadDeadline sets the read deadline
+func (q *ControlSocketConn) SetReadDeadline(t time.Time) error {
+	return errNotImplemented
+}
+
+// SetWriteDeadline sets the write deadline
+func (q *ControlSocketConn) SetWriteDeadline(t time.Time) error {
+	return errNotImplemented
+}
+
+// closeConn waits for the read end and the write end of th connection
+// to be closed and then closes the underlying connection
+func (q *ControlSocketConn) closeConn() {
+	// remember the ABC
+	<-q.readsCloseChan
+	<-q.writesCloseChan
+	q.cl.Close()
+}
+
+func makeControlSocketConn(conn io.ReadWriteCloser, fd uint32, cs *ControlSocket) *ControlSocketConn {
+	return &ControlSocketConn{
+		wb:              bufio.NewWriterSize(conn, buffSize),
+		rb:              bufio.NewReaderSize(conn, buffSize),
+		cl:              conn,
+		cb:              cs,
+		fd:              fd,
+		addr:            qemuAddr(""),
+		readsCloseChan:  make(chan struct{}),
+		writesCloseChan: make(chan struct{}),
+		readLock:        &sync.Mutex{},
+		writeLock:       &sync.Mutex{},
+		closeLock:       &sync.Mutex{},
+	}
+}
+
+type ControlSocket struct {
+	pipe      *Pipe
+	ml        sync.RWMutex
+	cond      *sync.Cond
+	m         map[uint32]*ControlSocketConn
+	available []uint32
+	c         net.Conn // Connection of control socket..
+}
+
+func (b *ControlSocket) Close() error {
+	return b.pipe.Close()
+}
+
+func (b *ControlSocket) close(q *ControlSocketConn) error {
+	b.ml.Lock()
+	delete(b.m, q.fd)
+	b.ml.Unlock()
+	if q.fd == 0 {
+		log.Print("Closing the control socket!")
+		//         b.c := nil
+		//         return nil
+	}
+
+	cl := control_socket.SocketControl_close
+	cinfo := &control_socket.SocketControl{
+		Fd:   &q.fd,
+		Sort: &cl,
+	}
+
+	out, err := proto.Marshal(cinfo)
+	if err != nil {
+		return err
+	}
+	_, err = b.c.Write(out)
+	return err
+}
+
+func (b *ControlSocket) Open(fd uint32) (net.Conn, error) {
+	cl := control_socket.SocketControl_identity
+	cinfo := &control_socket.SocketControl{
+		Fd:   &fd,
+		Sort: &cl,
+	}
+
+	base, err := b.pipe.Accept()
+	conn := makeControlSocketConn(base, fd, b)
+	go conn.closeConn()
+
+	if err != nil {
+		log.Printf("Failed to accept incoming connection %s", err)
+		return conn, err
+	}
+
+	out, err := proto.Marshal(cinfo)
+	if err != nil {
+		log.Printf("Failed to marshall identity announcement %s", err)
+		return conn, err
+	}
+
+	_, err = conn.Write(out)
+	if err != nil {
+		log.Printf("And it failed.")
+		return conn, err
+	}
+
+	// We are opening the control socket..
+	if fd == 0 {
+		b.c = conn
+		log.Printf("Openend the control socket %d", fd)
+		return conn, nil
+	}
+
+	c := makeControlSocketConn(conn, fd, b)
+	b.ml.Lock()
+	b.m[fd] = c
+	b.available = append(b.available, fd)
+	b.ml.Unlock()
+	b.cond.Signal()
+
+	return c, err
+}
+
+func (b *ControlSocket) nextConnection() (net.Conn, error) {
+	b.ml.Lock()
+	for len(b.available) == 0 {
```

Comment:
you can use a channel here:
   `id <- available`
it will block and wait for connection and remove the synchronization logic

Reply:
Ok. I'm not familiar with channels so I will read up on this and refactor.